go/graphd/io.go

```diff
@@ -0,0 +1,131 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The io portion of the graphd package facilitates IO operations against a graphd database.
+
+package graphd
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+)
+
+// readReponses returns a Response pointer slice read from an established connection to a graphd
+// database and a nil error on success.  On any failure to read a response, a Response pointer
+// slice containing zero-value Response pointers for the given failures, and the last encountered
+// error are returned.  readResponses must be called with g.conn locked.
+// TODO: Would it be useful to return all encountered errors?  Any chance after a failure that
+//       we'll block on the following call to ReadString()?
+func (g *graphd) readResponses(reqsNum int) ([]*Response, error) {
+	var retErr error
+	var resSlice []*Response
+
+	reader := bufio.NewReader(g.conn.netConn)
+
+	for i := 0; i < reqsNum; i++ {
+		str, err := reader.ReadString('\n')
+		if err != nil {
+			resSlice = append(resSlice, NewResponse(""))
+			retErr = err
+		} else {
+			resSlice = append(resSlice, NewResponse(str))
+		}
+	}
+
+	return resSlice, retErr
+}
+
+// Query attempts to send the Request to the graphd database to which this instance of the library
+// is connected.  In the case of more than one Request, the requests are joined and sent as one.
+// If no connection is currently present, or if the established connection is stale, Query will
+// trigger a Redial.  A Response pointer slice containing responses from the graphd database (failed
+// responses are zero-value value Response pointers) is returned along with an error code.  In the
+// case of any failed responses, the returned error code will contain the last encountered error.
+// Query locks the connection, allowing one thread to Query at a time.
+func (g *graphd) Query(reqs ...*Request) ([]*Response, error) {
+	g.conn.Lock()
+
+	// Join requests into one if needed.
+	var req *Request
+	reqsNum := len(reqs)
+	if reqsNum > 1 {
+		req = joinRequests(reqs...)
+	} else {
+		req = reqs[0]
+	}
+
+	g.LogDebugf("attempting to send '%v'", req)
+
+	sent := false
+	retries := 2
+	for sent == false {
+		var err error
+		var errStr string
+
+		switch g.conn.existsConnection() {
+		// An established connection is present, try to send.
+		case true:
+			// Queries to graphd are new line termianted.
+			_, err = fmt.Fprintf(g.conn.netConn, "%v", req.body)
+			if err != nil {
+				// Set base error for failed send.
+				errStr = fmt.Sprintf("failed to send '%v': %v", req, err)
+				retries--
+				// If we've exhausted our retries, log and return error.
+				if retries == 0 {
+					g.LogErr(errStr)
+					g.conn.Unlock()
```

Comment:
instead of sprinkling this around, defer g.conn.Unlock at the top?

Reply:
I think I ended up doing this in order to keep with the immediate error check convention.  I need to release the connection lock before the Redial attempt (line 92) and check for an error.  Doing the unlock as a defer, would lead me to:

```Go
g.conn.Unlock()
err = g.Redial(0)
g.conn.Lock()
if err != nil {
	errStr = fmt.Sprintf("%v: %v", errStr, err)
	g.LogErr(errStr)
	return []*Response{NewResponse("")}, errors.New(errStr)
}
```

If that doesn't hurt the eyes too much, I can get rid of all the explicit connection unlocks and go with a defer.